File: src/hexd.cpp
/**
 * Tool for displaying file contents in hexadecimal, octal,
 * binary and ascii formats.
 * Copyright (C) 2024 Alexander (@alkuzin).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <hexd.hpp>
#include <fstream>
#include <cstring>
#include <print>


namespace hexd {

inline const auto HEXD_VERSION  {"2.0"};
inline const auto HEXD_YEAR     {"2024"};
inline const auto FMT_SIZE      {8};
inline const auto HEXD_TITLE {
    "\n"
    " ██╗  ██╗███████╗██╗  ██╗██████╗  \n"
    " ██║  ██║██╔════╝╚██╗██╔╝██╔══██╗ \n"
    " ███████║█████╗   ╚███╔╝ ██║  ██║ \n"
    " ██╔══██║██╔══╝   ██╔██╗ ██║  ██║ \n"
    " ██║  ██║███████╗██╔╝ ██╗██████╔╝ \n"
    " ╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝╚═════╝  \n"
};

void help(void)
{
    std::println(
        "{}\nUSAGE\n"
        "\thexd [options] <file>\n\n"
        "DESCRIPTION\n"
        "\thexd - display file contents in hexadecimal, octal, "
        "binary and ascii formats\n\n"
        "OPTIONS\n"
        "\t-x\t\tone-byte hexidecimal+ASCII display\n"
        "\t-o\t\tone-byte octal display\n"
        "\t-b\t\tone-byte binary display\n\n"
        "\t-h, --help\tdisplay options list\n"
        "\t-v, --version\tdisplay version of hexd", HEXD_TITLE
    );
    std::exit(EXIT_SUCCESS);
}

void version(void)
{
    std::println("{}\nhexd: v{} {} (@alkuzin)", HEXD_TITLE, HEXD_VERSION, HEXD_YEAR);
    std::exit(EXIT_SUCCESS);
}

/**
 * @brief Convert byte to ASCII.
 *
 * @param [in] ch - given byte to convert.
 * @return character to print.
 */
inline char to_print(uint8_t ch) noexcept
{
    return (ch > 31 && ch < 127)? ch : '.';
}

/**
 * @brief Handle bytes output mode.
 *
 * @param [in] mode - given bytes output mode.
 * @param [out] bpl - given bytes per line.
 * @param [out] fmt - given format string.
 */
static void handle_mode(MODE mode, std::size_t *bpl, std::string_view& fmt) noexcept
{
    switch(mode) {
		case MODE::BIN:
            *bpl = 8;
            fmt = "%08b ";
			break;

		case MODE::OCT:
            *bpl = 16;
            fmt = "%03o ";
			break;

		case MODE::HEX:
            *bpl = 16;
            fmt = "%02x ";
			break;

		default:
            *bpl = 16;
            fmt = "%02x ";
			break;
	};
}

void dump(MODE mode, const std::string_view& filename) noexcept
{
    std::ifstream fin(filename.data(), std::ios::in);

    if (!fin.is_open()) {
        std::println("error to open file \"{}\"", filename);
        std::exit(EXIT_FAILURE);
    }

    std::size_t      bytes_per_line {16};
    std::string_view fmt;

    handle_mode(mode, &bytes_per_line, fmt);

    std::uint8_t buffer[bytes_per_line];
    std::size_t  rows = 0;

    while (fin) {
        std::memset(buffer, 0, sizeof(buffer));
        fin.read(reinterpret_cast<char*>(buffer), bytes_per_line);

        std::print("{:08x}  ", rows);

        // print first half of 8 bytes in hexadecimal format
        for (std::size_t i = 0; i < bytes_per_line >> 1; i++)
            std::printf(fmt.data(), buffer[i]);

        std::putchar(' ');

        // print second half of 8 bytes in hexadecimal format
        for (std::size_t i = bytes_per_line >> 1; i < bytes_per_line; i++)
            std::printf(fmt.data(), buffer[i]);

        // print string representation of bytes
        std::putchar(' ');
        std::putchar('|');

        for (const auto& byte : buffer)
            std::putchar(to_print(byte));

        std::putchar('|');
        std::putchar('\n');

        rows += 0x10;
    }

    fin.close();
}

} // namespace hexd

File: src/hexd.rs
// Tool for displaying file contents in hexadecimal, octal,
// binary and ascii formats.
// Copyright (C) 2025 Alexander (@alkuzin).
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! File dump functions.

use std::io::{BufReader, Read, Write};
use std::fs::File;

/// Output mode enumeration
#[derive(Debug)]
pub enum Mode {
    Hex,
    Oct,
    Bin,
}

/// Convert byte to ASCII.
///
/// # Parameters
/// - `ch` - given byte to convert.
///
/// # Returns
/// ASCII character to print.
#[inline]
fn to_print(ch: u8) -> char {
    if ch.is_ascii_graphic() {ch as char} else {'.'}
}

/// Get bytes per line for specific output mode.
///
/// # Parameters
/// - `mode` - given output mode.
///
/// # Returns
/// Bytes per line.
fn get_bytes_per_line(mode: &Mode) -> usize {
    match mode {
        Mode::Hex => return 16,
        Mode::Oct => return 16,
        Mode::Bin => return 8,
    }
}

/// Dump file.
///
/// # Parameters
/// - `mode` - given output mode.
/// - `filename` - given name of file to dump.
pub fn dump(mode: &Mode, filename: &str) -> Result<(), std::io::Error> {
    let file       = File::open(filename).expect("hexd: error to open file");
    let mut reader = BufReader::new(file);

    let bytes_per_line      = get_bytes_per_line(mode);
    let mut buffer: Vec<u8> = vec![0; bytes_per_line];
    let mut rows: usize     = 0;
    let mut output          = String::with_capacity(96);

    loop {
        let bytes_read = reader.read(&mut buffer)?;

        if bytes_read == 0 {
            break;
        }

        output.clear();
        output.push_str(&format!("{:08x}  ", rows));

        for i in 0..bytes_per_line {
            match mode {
                Mode::Hex => output.push_str(&format!("{:02x} ", buffer[i])),
                Mode::Oct => output.push_str(&format!("{:03o} ", buffer[i])),
                Mode::Bin => output.push_str(&format!("{:08b} ", buffer[i])),
            }

            if (i + 1) % 8 == 0 {
                output.push_str(" ");
            }
        }

        output.push_str("|");

        for i in 0..bytes_per_line {
            output.push_str(&format!("{}", to_print(buffer[i])));
        }

        if let Err(e) = writeln!(std::io::stdout(), "{}|", output) {
            // Ignore broken pipe error
            if e.kind() == std::io::ErrorKind::BrokenPipe {
                return Ok(());
            }
            else {
                return Err(e);
            }
        }

        rows += bytes_per_line;
        buffer.fill(0);
    }

    Ok(())
}